src/Storages/FileLog/StorageFileLog.cpp

```diff
@@ -0,0 +1,746 @@
+#include <Interpreters/evaluateConstantExpression.h>
+#include <Parsers/ASTCreateQuery.h>
+#include <Parsers/ASTExpressionList.h>
+#include <Parsers/ASTInsertQuery.h>
+#include <Parsers/ASTLiteral.h>
+#include <Processors/Executors/CompletedPipelineExecutor.h>
+#include <Processors/Pipe.h>
+#include <Processors/Transforms/ExpressionTransform.h>
+#include <Storages/FileLog/FileLogSource.h>
+#include <Storages/FileLog/ReadBufferFromFileLog.h>
+#include <Storages/FileLog/StorageFileLog.h>
+#include <Storages/StorageFactory.h>
+#include <Storages/StorageMaterializedView.h>
+#include <Poco/File.h>
+#include <Common/Exception.h>
+#include <Common/Macros.h>
+#include <Common/getNumberOfPhysicalCPUCores.h>
+#include <Common/quoteString.h>
+#include <Common/typeid_cast.h>
+#include <common/logger_useful.h>
+
+#include <sys/stat.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int BAD_ARGUMENTS;
+    extern const int CANNOT_GET_FILE_STAT;
+    extern const int NOT_REGULAR_FILE;
+    extern const int READ_META_FILE_FAILED;
+    extern const int FILE_STREAM_ERROR;
+    extern const int LOGICAL_ERROR;
+}
+
+namespace
+{
+    const auto RESCHEDULE_MS = 500;
+    const auto BACKOFF_TRESHOLD = 32000;
+    const auto MAX_THREAD_WORK_DURATION_MS = 60000;
+}
+
+StorageFileLog::StorageFileLog(
+    const StorageID & table_id_,
+    ContextPtr context_,
+    const ColumnsDescription & columns_,
+    const String & relative_path_,
+    const String & format_name_,
+    std::unique_ptr<FileLogSettings> settings,
+    bool attach)
+    : IStorage(table_id_)
+    , WithContext(context_->getGlobalContext())
+    , filelog_settings(std::move(settings))
+    , path(getContext()->getUserFilesPath() + "/" + relative_path_)
+    , format_name(format_name_)
+    , log(&Poco::Logger::get("StorageFileLog (" + table_id_.table_name + ")"))
+    , milliseconds_to_wait(RESCHEDULE_MS)
+{
+    StorageInMemoryMetadata storage_metadata;
+    storage_metadata.setColumns(columns_);
+    setInMemoryMetadata(storage_metadata);
+
+    try
+    {
+        loadMetaFiles(attach);
+        loadFiles();
+
+        assert(file_infos.file_names.size() == file_infos.meta_by_inode.size());
+        assert(file_infos.file_names.size() == file_infos.inode_by_name.size());
+        assert(file_infos.file_names.size() == file_infos.context_by_name.size());
+
+        if (path_is_directory)
+            directory_watch = std::make_unique<FileLogDirectoryWatcher>(path);
+
+        auto thread = getContext()->getMessageBrokerSchedulePool().createTask(log->name(), [this] { threadFunc(); });
+        task = std::make_shared<TaskContext>(std::move(thread));
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+}
+
+void StorageFileLog::loadMetaFiles(bool attach)
+{
+    const auto database = DatabaseCatalog::instance().getDatabase(getStorageID().getDatabaseName());
```

Comment:
It works for attach table, but seems not work for attach partial dropped table. I don't find a better way to do this.

Reply:
>  but seems not work for attach partial dropped table

what does it mean `attach partial drop table`?

> It works for attach table

Simple detach-attach table can differ from server restart, did you check restart?

Reply:
Yes, I checked restart.

Reply:
> > but seems not work for attach partial dropped table
> 
> what does it mean `attach partial drop table`?
> 

When drop table, the table meta file(.sql) will first moved to `metadata_dropped`, then clear in background, if server restart before clear, it will try to attach the partial dropped table.

Reply:
https://github.com/ClickHouse/ClickHouse/blob/2805c28e6573fab09128b43150c4a7cb7ad21cc1/src/Interpreters/DatabaseCatalog.cpp#L681-L689

Reply:
Then may be we can use path to table data for meta also:
https://github.com/ClickHouse/ClickHouse/blob/2805c28e6573fab09128b43150c4a7cb7ad21cc1/src/Storages/StorageFactory.h#L39-L40

We can get it from args in registerStorage.